### ISSUE TITLE
change Dependency.OnValue<T>(object value) to (T Value)

### ISSUE THREAD

**Issue by abbotware** [castleproject/Windsor#465]
currently, this overload of OnValue has this definition:

`public static Property OnValue<TDependencyType>(object value);`

wouldn't

`public static Property OnValue<TDependencyType>(`**TDependencyType**` value);`

make more sense? (and catch more bugs via the compiler)

**Comment by matgrioni** [comment]
Also ran into this issue recently when changing types in a constructor and forgetting these values were supplied via DI. Seems like it should be relatively straightforward, not sure what might might be missing.

**Comment by jonorossi** [comment]
It might have been done this way on purpose to force specifying the generic type parameter since linting tools like ReSharper (and now Roslyn) suggest to remove redundant generic parameters.

If you had the following, and removed the generic parameter because it is redundant, then changed the `thing` variable to `var` it would pass a different type to Windsor:

```csharp
IThing thing = new Thing();
Container.Register(
    Component.For<X>().DependsOn(Dependency.OnValue<IThing>(thing))
);
```

Looks like a tradeoff both ways.

**Comment by matgrioni** [comment]
That a good point, since always supplying it would mean the type it supplied for DI would always be obvious.

Would a runtime check in the OnValue method be a reasonable addition to make sure the value is always a subtype of the provided generic type?

**Comment by jonorossi** [comment]
> Would a runtime check in the OnValue method be a reasonable addition to make sure the value is always a subtype of the provided generic type?

Sounds like a reasonable addition, did you want to submit a pull request?

**Comment by abbotware** [comment]
I think we are conflating two separate things:


1
```
IThing thing = new Thing();
Container.Register(
    Component.For<X>().DependsOn(Dependency.OnValue<IThing>(thing))
);
```

vs

2
```
var thing = new NotAThing();
Container.Register(
    Component.For<X>().DependsOn(Dependency.OnValue<IThing>(thing))
);
```

the second scenario compiles now!  I am proposing a changing this:
`public static Property OnValue<TDependencyType>(TDependencyType value);`

which would prevent 2 from even compiling

**Comment by abbotware** [comment]
The problem with making this change as is in place  that it will more than likely break a lot of code... which is bad for package maintainers - instead the existing needs to be marked obsolete and 2 new methods: should be created:

```
[Obsolete('use OnObjectValue or OnTypedValue<> instead")]
public static Property OnValue<TDependencyType>(object value);

public static Property OnObjectValue (object value);
// This is the old method for those that really want to use 'object'
// note the lack of generics since it served no purpose in the first place

public static Property OnTypedValue<TDependencyType>(TDependencyType value);
// This is the new / better version with type safety
```

**Comment by jonorossi** [comment]
@abbotware I'm not sure you understood what I described above. I've repeated it with some context.

> If you had the following (and `OnValue` accepted `TDependencyType` rather than `object`):

```csharp
IThing thing = new Thing();
Container.Register(
    Component.For<X>().DependsOn(Dependency.OnValue<IThing>(thing))
);
```

> and removed the generic parameter because it is redundant:

![Screen Shot 2019-08-18 at 11 07 31 pm](https://user-images.githubusercontent.com/126297/63224830-302b0580-c20d-11e9-933a-52b1e0e15fcd.png)

> then changed the thing variable to var it would pass a different type to Windsor

![Screen Shot 2019-08-18 at 11 08 40 pm](https://user-images.githubusercontent.com/126297/63224832-37eaaa00-c20d-11e9-8b1c-93200f7f994e.png)

I'm not saying this is all done in the same step, but refactoring this code using linting tools like ReSharper and Roslyn will break your code.

**Comment by abbotware** [comment]
I understand completely - I create read only/immutable interfaces for my objects and inject those into the container.    

Refactoring tools are not a panacea - I believe it is even trying to warn you that something might be undesirable since the warning **inferred type will change** appears - I use the var = refactoring all the time and I have never seem that message in VS or Resharper!  I would wager that warning is not present in in other scenarios for that refactoring feature when it doesn't change the type..  so looks like user error :-)   

I haven't verified it, but I think the solution I proposed might actually _prevent_ refactoring tools from making a mistake since it is more strongly typed.  Right now, anything you pass to OnValue is only an **Object** 

if **OnValue**  expected a parameter of **TDependencyType** instead, it would preserve type information as most FluentAPI's do when properly implemented.

**Comment by jonorossi** [comment]
> public static Property OnObjectValue (object value);
> // This is the old method for those that really want to use 'object'
> // note the lack of generics since it served no purpose in the first place

@abbotware what do you mean by the generic type served no purpose? It is the key:

```csharp
public static Property OnValue<TDependencyType>(object value)
{
	return Property.ForKey<TDependencyType>().Eq(value);
}
```

**Comment by abbotware** [comment]
whoops. I forgot about that!

In either case - it's not used to enforce the type of 'object value'  It seems like such a trivial change - however, it might have many unintended consequences - Code that compiles today, might stop compiling once this is changed.

Hence the reason recommend to mark it deprecated, and introduce 2 new versions.  

Or just provide provide the OnTypedValue variant which is Strongly Typed and hope the old one falls out of favor

**Comment by jonorossi** [comment]
This has gone stale so I'm going to close it. I still think the unintended consequences are more important here than the compile time check. We could go with a runtime check in the future, or maybe even a chained API where the generic type parameter can't inferred so linting tools can't remove it being specified.

**Comment by abbotware** [comment]
How is a compile time check not better?

**Comment by jonorossi** [comment]
> How is a compile time check not better?

I've explained the unintended consequences multiple times in this issue, please reread the issue if you don't recall. I even proposed an idea for a compile time check which would avoid those unintended consequences in my last comment.

**Comment by abbotware** [comment]
So why close this issue then?

**Comment by abbotware** [comment]
Your idea makes no sense as stated,

> maybe even a chained API where the generic type parameter can't inferred so linting tools can't remove it being specified.

**NEW PROPOSAL**

```
public static Property OnTypedValue<TDependencyType, TValueType>(TValueType value)
     where TValueType : TDependencyType
   {
     return Property.OnValue<TDependencyType>(value);
}
```

With 2 type parameters, that are related via inheritance, I doubt any linting tool would recommend removal

**Comment by jonorossi** [comment]
> Your idea makes no sense as stated,
> 
> > maybe even a chained API where the generic type parameter can't inferred so linting tools can't remove it being specified.

`Dependency.OnValue` is implemented as a single line chained method call. `Property.ForKey<>` returns a `PropertyKey` and `Eq` returns a `Property`. My suggestion was pushing the compiler safety into the `Property` class, i.e. `Property.ForKey<TKey>` returning a `PropertyKey<TKey>` so `PropertyKey.Eq` would become generic and only accept the specified type.

> **NEW PROPOSAL**
> 
> ...
> 
> With 2 type parameters, that are related via inheritance, I doubt any linting tool would recommend removal

Your new proposal could work, but doesn't that mean you always have to specify the generic type parameter twice?

**Comment by abbotware** [comment]
> your new proposal could work, but doesn't that mean you always have to specify the generic type parameter twice?

yes, but that is the entire point of using this sort of technique

Specifying the types explicitly just makes it extremely obvious that something special is happening when using OnValue/Dependency notation. Without it can be error prone due to the 'object' in the return - (hence this issue being opened in the first place)  

With this overload there would be zero chance this would be used incorrectly, by accident, or change existing behavior.   Even in my PR I renamed the function so it would have broken anything, but with 2 Type parameters (I think I prefer this now), it can be keep the same name.  I think with 2 types, the compiler won't guess or try to infer the type

**Comment by jonorossi** [comment]
@abbotware any opinion about the first half of my comment...?

We've already got:
```csharp
Property.ForKey<TKey>().Eq(value)
```

If we changed `ForKey<TKey>()` to return a new class `PropertyKey<TKey>` that only handled `Type` keys (and not `string` keys) you'd get the compiler type checking without double specifying the generic type parameter, and probably also get compiler type checking for the `Is<T>` method for service overrides at the same time.

**Comment by eugbaranov** [comment]
> [@abbotware](https://github.com/abbotware) any opinion about the first half of my comment...?
> 
> We've already got:
> 
> Property.ForKey<TKey>().Eq(value)
> If we changed `ForKey<TKey>()` to return a new class `PropertyKey<TKey>` that only handled `Type` keys (and not `string` keys) you'd get the compiler type checking without double specifying the generic type parameter, and probably also get compiler type checking for the `Is<T>` method for service overrides at the same time.

For visibility, I would suggest adding an overload to `Dependency.OnValue` to return typed `Property` so it can be used like:

    Dependency.OnValue<TKey>().Eq(value)
